tmt/steps/provision/artemis.py

```diff
@@ -0,0 +1,635 @@
+        self.api_retry_backoff_factor = data['api-retry-backoff-factor']
+
+    def save(self):
+        data = cast(GuestDataType, super().save())
+
+        data['api-url'] = self.api_url
+        data['api-version'] = self.api_version
+        data['arch'] = self.arch
+        data['image'] = self.image
+        data['hardware'] = self.hardware
+        data['pool'] = self.pool
+        data['priority-group'] = self.priority_group
+        data['keyname'] = self.keyname
+        data['user-data'] = self.user_data
+        data['guestname'] = self.guestname
+        data['guest'] = self.guest
+        data['user'] = self.user
+        data['provision-timeout'] = self.provision_timeout
+        data['provision-tick'] = self.provision_tick
+        data['api-timeout'] = self.api_timeout
+        data['api-retries'] = self.api_retries
+        data['api-retry-backoff-factor'] = self.api_retry_backoff_factor
+
+        return data
+
+    def _create(self) -> None:
+        environment: Dict[str, Any] = {
+            'hw': {
+                'arch': self.arch
+                },
+            'os': {
+                'compose': self.image
+                }
+            }
+
+        data: Dict[str, Any] = {
+            'environment': environment,
+            'keyname': self.keyname,
+            'priority_group': self.priority_group,
+            'user_data': self.user_data
+            }
+
+        if self.pool:
+            environment['pool'] = self.pool
+
+        if self.hardware is not None:
+            assert isinstance(self.hardware, dict)
+
+            environment['hw']['constraints'] = self.hardware
+
+        # TODO: snapshots
+        # TODO: spot instance
+        # TODO: post-install script
+        # TODO: log types
+
+        response = self.api.create('/guests/', data)
+
+        if response.status_code == 201:
+            self.info('guest', 'has been requested', 'green')
+
+        else:
+            raise ProvisionError(
+                f'failed to create, unhandled API response {response.status_code}')
+
+        self.guestname = response.json()['guestname']
+
+        self.info('guestname', self.guestname, 'green')
+
+        deadline = datetime.datetime.utcnow(
+            ) + datetime.timedelta(seconds=self.provision_timeout)
+
+        while deadline > datetime.datetime.utcnow():
+            response = self.api.inspect(f'/guests/{self.guestname}')
+
+            if response.status_code != 200:
+                raise ProvisionError(
+                    f'failed to create, unhandled API response {response.status_code}')
+
+            current = cast(GuestInspectType, response.json())
+
+            state = current['state']
+
+            if state == 'error':
+                raise ProvisionError(f'failed to create, provisioning failed')
+
+            if state == 'ready':
+                self.guest = current['address']
+
+                self.info('address', self.guest, 'green')
+
+                break
```

Comment:
```suggestion
                return
```

I guess you meant `return` here?

Reply:
But then the `failed to provision in...` error is always raised...

Reply:
Ok, that's an actual bug, I'll fix it right away.

Reply:
Fixed.

Reply:
Thanks for the quick fix.